### ISSUE TITLE
C-API has memory leaks

### ISSUE THREAD

**Issue by zenna** [dreal/dreal3#116]
From dReal.jl (this file is now in tests)

```julia
ctx = Context(qf_nra)
X = Var(ctx, Float64,"X", 0.0,1.0)
for i = 1:10000
  push_ctx!(ctx)
  a = rand()
  b = rand()
  add!(ctx,(<=)(ctx,X,b))
  add!(ctx,(>=)(ctx,X,a))
  is_satisfiable(ctx)
  pop_ctx!(ctx)
end
```

**Comment by soonhokong** [comment]
@zenna, this is a known issue. I've assigned a lower priority so far. I agree that it's an important one to fix now.

**Comment by zenna** [comment]
Ok @soonhokong . Do you know Is it a problem in the C-API or the C++?

**Comment by soonhokong** [comment]
It should be on C++ side.

**Comment by soonhokong** [comment]
I've fixed some problems causing memory leaks. I've found some problems on IBEX. I'll report them to IBEX team. Thanks for your patience.

**Comment by zenna** [comment]
Great.  Let me know what were/are the IBEX issues, because now I am using IBEX directly.

**Comment by soonhokong** [comment]
I opened an issue, https://github.com/ibex-team/ibex-lib/issues/137.

**Comment by zenna** [comment]
Any word on this? I saw you made changes so recompiled dReal but the above code leaks memory (extremely rapidly).

**Comment by soonhokong** [comment]
Well.. I'm still fixing other leaks. I will let you know when it's ready.

BTW, how can I check memory leaks in Julia? Could you teach me how to do
that? For C/C++ code, I'm using valgrind.

>

-- 

Best,
Soonho

**Comment by zenna** [comment]
I found the memory leaks simply because my system ran out of memory.  But you can use valgrind with Julia:

http://julia.readthedocs.org/en/latest/devdocs/valgrind/

**Comment by soonhokong** [comment]
@zenna, I see. Thanks for the link!

**Comment by soonhokong** [comment]
@zenna, I just made another commit https://github.com/dreal/dreal3/commit/788e94b7ef0c7f3f3bb906c3721f2423a7e75167. It should fix the problem, could you test it?

Also, if you're using ibex on your own, Reading the commit message of 788e94b7ef0c7f3f3bb906c3721f2423a7e75167 should be useful for you to avoid this kind of memory leak. It was pretty much difficult for me to catch.

**Comment by zenna** [comment]
@soonhokong Testing now.

That is a weird bug.

**Comment by zenna** [comment]
@soonhokong The same problem remains.

I created the following program directly in C

```C
// Test Memory Leaks
int main(int argc, char * argv[]) {
    int i = 0;
    opensmt_init();
    opensmt_context ctx = opensmt_mk_context(qf_nra);
    opensmt_set_precision (ctx, 0.0000001);
    opensmt_expr x = opensmt_mk_real_var(ctx, "x" , 0.0, 1.0);
    for (i=0;i<10000;++i) {
        opensmt_push(ctx);
        opensmt_expr lb = opensmt_mk_num_from_string(ctx, "0.1");
        opensmt_expr ub = opensmt_mk_num_from_string(ctx, "0.9");
        opensmt_expr geq = opensmt_mk_geq(ctx, x, lb);
        opensmt_expr leq = opensmt_mk_leq(ctx, x, lb);
        opensmt_assert(ctx, geq);
        opensmt_assert(ctx, leq);
        opensmt_result res = opensmt_check( ctx );
        printf( "%s\n\n", res == l_false ? "unsat" : "sat" );
        opensmt_pop(ctx);
    }
    opensmt_del_context(ctx);
    return 0;
}
```

And even using the old dReal libraries, the memory didn't explode, although one would have to check with valgrind to see if the leaks are actually gone.

This means there's something wrong with the Julia interface, I'll try to figure that out.

**Comment by soonhokong** [comment]
@zenna, thanks for testing it.

> one would have to check with valgrind to see if the leaks are actually gone.

I'll check this. If there is any remaining leaks from this C code, I'll fix them.

**Comment by soonhokong** [comment]
I just ran valgrind. It shows that there is no leak, however the 'zero' variable that I introduced in https://github.com/dreal/dreal3/commit/788e94b7ef0c7f3f3bb906c3721f2423a7e75167 is *still reachable*:

This variable is *static*, that is, it will be created only once. The size of wasted memory should be less than 256 bytes. Anyway, I will revise the code to free this variable in the end.

**Comment by soonhokong** [comment]
BTW, I just take your code and add it to `src/tests/c_api/leak_test.c`. After fetching up the latest version of source code and recompiling them, you can run the leak_test by executing `./tests/c_api/c_api_leak_test`. You can also run valgrind over it by executing `valgrind --leak-check=full ./tests/c_api/c_api_leak_test`. When you get the latest version, please try valgrind and let me know if you have a different result.

You might have about 70K of *still reachable* region. This should be fine based on http://stackoverflow.com/questions/30376601/valgrind-memory-still-reachable-with-trivial-program-using-iostream.

```
==27671== 72,704 bytes in 1 blocks are still reachable in loss record 1 of 1
==27671==    at 0x4C2AB9D: malloc (vg_replace_malloc.c:296)
==27671==    by 0x4EC060F: ??? (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21)
==27671==    by 0x400F305: call_init.part.0 (dl-init.c:85)
==27671==    by 0x400F3DE: call_init (dl-init.c:52)
==27671==    by 0x400F3DE: _dl_init (dl-init.c:134)
==27671==    by 0x40016E9: ??? (in /lib/x86_64-linux-gnu/ld-2.15.so)
==27671== 
```

**Comment by zenna** [comment]
Yeah

Something strange is going on at the Julia side.  There is a small minor difference between the C leak test and the julia leak test which is that the julia leak test had constraints of the form

```
b = rand()
add!(ctx,(<=)(ctx,X,b))
```

But the c form had `b` hard coded as 0.1.  When I use the hard coded values in Julia there is no memory leak, otherwise it eats up about 1 gigabyte per second!

When I make the C code use randomly generated values, there is still no memory leak.

So maybe the memory for each of these greater-than and less-than constraints is not being freed by `pop` at the end of each iteration?  But this still doesn't explain the disparity between the two cases.

I'm currently recompiling Julia with these valgrind options installed to see if that will help.

**Comment by soonhokong** [comment]
hmm... this sounds very strange. Please let me know when you find something.

**Comment by soonhokong** [comment]
Possibly, it's because of the way we construct a constant. So far, dReal only provide `opensmt_mk_num_from_string` function which takes a string representation of a number. I can add `opensmt_mk_num(opensmt_context, double const)` which directly takes a double. Let's see how it helps.

**Comment by zenna** [comment]
Perhaps.  Let me know when once you've added it.

**Comment by soonhokong** [comment]
>  I can add `opensmt_mk_num(opensmt_context, double const)` which directly takes a double

It's added by 387cc390dccb9e4a4ec0df27dac4f85b7a99a6bb.

**Comment by soonhokong** [comment]
I also chaged `leak_test.c` to use `rand()` function. See https://github.com/dreal/dreal3/commit/ccaaeb8be074c8a1d47b9a73c75be19d46f31b16 for details.

**Comment by soonhokong** [comment]
Let me fix the *still reacahable* issue soon. I want to make sure that we don't waste any memory in dReal side.

**Comment by zenna** [comment]
@soonhokong I don't know whether it is the result of this change or the fact that I cloned a fresh dReal and built from scratch, but the problem seems to have gone!

There is still an increase in memory, over time, but its only about 1 mg per second (instead of 1 gigabyte) and it could well be just down to the fact that Julia is memory managed and the garbage collector is yet to kick in.

**Comment by soonhokong** [comment]
That's great! BTW I just submitted another commit to the repo, please try it when you have time. Now, I can't find any *leak*/*still reachable*.
